core/src/main/java/com/linecorp/armeria/common/stream/AbstractStreamMessageDuplicator.java

```diff
@@ -587,20 +514,175 @@ private void accumulateDemand(long n) {
+                    return;
+                }
+
+                if (offset == endOffset) {
+                    break;
+                }
+
+                if (!doSignalSingle(signals)) {
+                    break;
+                }
+            }
+        }
+
+        private boolean doSignalSingle(List<Object> signals) {
+            final Object signal = signals.get(offset);
+            if (signal instanceof CloseEvent) {
+                if (!beginSignal()) {
+                    return false;
+                }
+
+                try {
+                    offset++;
+                    processor.unsubscribe(this, ((CloseEvent) signal).cause);
+                    return true;
+                } finally {
+                    endSignal();
+                }
+            }
+
+            for (;;) {
+                final long demand = this.demand;
+                if (demand == 0) {
+                    break;
+                }
+
+                if (demand != Long.MAX_VALUE && !demandUpdater.compareAndSet(this, demand, demand - 1)) {
+                    // Failed to decrement the demand due to contention.
+                    continue;
+                }
+
+                if (!beginSignal()) {
+                    // Succeeded to decrement the demand but failed to start the signal; increment back.
+                    incrementDemand();
+                    break;
+                }
+
+                try {
+                    offset++;
+                    @SuppressWarnings("unchecked")
+                    T obj = (T) signal;
+                    ReferenceCountUtil.touch(obj);
+                    if (withPooledObjects) {
+                        if (obj instanceof ByteBufHolder) {
+                            obj = retainedDuplicate((ByteBufHolder) obj);
+                        } else if (obj instanceof ByteBuf) {
+                            obj = retainedDuplicate((ByteBuf) obj);
+                        }
+                    } else {
+                        if (obj instanceof ByteBufHolder) {
+                            obj = copy((ByteBufHolder) obj);
+                        } else if (obj instanceof ByteBuf) {
+                            obj = copy((ByteBuf) obj);
+                        }
+                    }
+
+                    publishedAny = true;
+                    subscriber.onNext(obj);
+                    return true;
+                } finally {
+                    endSignal();
+                }
+            }
+
+            return false;
+        }
+
+        private void incrementDemand() {
+            for (;;) {
+                final long oldDemand = demand;
+                if (oldDemand == Long.MAX_VALUE) {
+                    break;
+                }
+
+                if (demandUpdater.compareAndSet(this, oldDemand, oldDemand + 1)) {
+                    break;
+                }
+            }
+        }
+
+        private boolean beginSignal() {
+            if (signaling) {
```

Comment:
I'm not sure, but doesn't signaling need to be volatile? signal() can be called from different two threads which are the writing thread to upstream and the subscribing thread who calls subscribe() -> onSubscribe() -> request().. 
Did I miss something? :-)